### ISSUE TITLE
Add more `t.Parallel()` to speed up tests

### ISSUE THREAD

**Issue by machine424** [prometheus/prometheus#15185]
### Proposal

Let's save some time.

- The packages/subdirectories to be checked:
  - [ ] `cmd/prometheus`, see https://github.com/prometheus/prometheus/pull/14618
  - [ ] `cmd/promtool`
  - [ ] `config/`
  - [ ] `discovery/`
  - [ ] `mode/`
  - [ ] `notifier/`
  - [ ] `rules/`
  - [ ] `scrape/`
  - [ ] `storage/`
  - [ ] `tsdb/`
  - [ ] `util/`
  - [ ] `web/`
  - [ ] `tracing/`
  
  
Note that:
- Some tests may not be designed to run in parallel.
- We should avoid sacrificing test readability (correctness a well of course :)) to make them runnable in parallel.
- Check if subtests can be run in parallel as well.
- If a test runs subtests unofficially (without t.Run()), consider adding the needed wrappers to run the subtests in parallel.
- Please make sure to run the tests locally with the race detector enabled (-race) as it's temporarily disabled in the CI.
- Of course, only make the test run in parallel if it would reduce its execution time.
- Let's consider the wall-clock time/execution time of the test in the CI as a key reference point for measuring improvements.

Tests can always be run sequentially using`-parallel 1`.

**Comment by Manask322** [comment]
@machine424  - can I take up adding testing cases in util and increasing the coverage. Will add parallel test cases as well whenever I can

**Comment by machine424** [comment]
> @machine424 - can I take up adding testing cases in util and increasing the coverage. Will add parallel test cases as well whenever I can

Sure thing, thanks.

**Comment by Manask322** [comment]
@machine424  - will it be possible to check this PR - https://github.com/prometheus/prometheus/pull/15188. have added for one of the util package. Since it is my first contribution wanted to keep the PR small for review and any comments. Will add for other utils once the above is merged. Thanks !!

**Comment by aniketnk** [comment]
I'm taking up `cmd/promtool`.

**Comment by aniketnk** [comment]
IMO, we should target speeding up the tests that have a longer runtime in CI. 
I opened one PR for `discovery/kubernetes`:#15227 and another for `cmd/promtool`: #15226.

PR https://github.com/prometheus/prometheus/pull/15227 reduces test execution time from ~34s to ~6s. 
Whereas https://github.com/prometheus/prometheus/pull/15226 makes almost no impact on execution time since the parallel testcases are small. 

@machine424 Please review both. 
Thanks :)

**Comment by 3Juhwan** [comment]
May I try `web/`?

**Comment by Leanndrosouza** [comment]
I'm taking up the `scrape/` package

**Comment by Leanndrosouza** [comment]
@machine424 please review #15246 when you can, thanks in advance

**Comment by 3Juhwan** [comment]
Hi @machine424, could you take a look at this PR #15248 when you can? I’ve made some optimizations to reduce test execution time, and your feedback would be greatly appreciated! 😄

**Comment by akulabs8** [comment]
Hey, I am new here and looking to contribute, so can I chip in here with any help?

**Comment by Vandit1604** [comment]
I'm picking up `config/`.

**Comment by jvalecillos** [comment]
I've created a PR for the `models` package: #5349

**Comment by Vandit1604** [comment]
@machine424 for the `config` package on main https://github.com/prometheus/prometheus/commit/fc3d19725d83e13363ce40d56672c6f97e2955e7
```sh
❯ go test ./config -count=1
ok  	github.com/prometheus/prometheus/config	0.074s
```

with parallelization, which could only be applied in [config/config_test.go](https://github.com/prometheus/prometheus/blob/main/config/config_test.go)
```sh
❯ go test ./config -count=1
ok  	github.com/prometheus/prometheus/config	0.078s
```

the change in execution time is not significant. Should I create a PR for this?

**Comment by Vandit1604** [comment]
#15367
I've raised a PR for `prompb` package

**Comment by sammyqtran** [comment]
Hi, I’d like to pick up where 3Juhwan left off on speeding up tests in the /web package. I’m working on addressing the maintainer’s comments on the existing PR [#15248](https://github.com/prometheus/prometheus/pull/15248) and plan to open a new PR with those fixes soon. Please let me know if that’s okay.

**Comment by odlev** [comment]
I'd like to work on the storage/remote package - client_test.go and codec_test.go look parallelizable.

**Comment by ogulcanaydogan** [comment]
Picking up `discovery/consul` — all 13 top-level test functions use independently created `httptest.Server` and `prometheus.Registry` instances, so there is no shared global state. PR incoming.

**Comment by ogulcanaydogan** [comment]
Adding `t.Parallel()` to `discovery/marathon` (13 tests). Same pattern as #18611 (consul) — no shared mutable state, safe to parallelize. PR coming shortly.

**Comment by ogulcanaydogan** [comment]
Will add `t.Parallel()` to `discovery/nomad` — same pattern as the openstack and puppetdb PRs.

**Comment by s23asd321-del** [comment]
Hi, I'd like to take a small, scoped part of this issue: parallelizing the independent subtests in `web/api/v1/errors_test.go`'s `TestApiStatusCodes`.\n\nI checked the current open PRs and did not find one covering this file/test. My local baseline shows `TestApiStatusCodes` is currently the slowest top-level test in `web/api/v1` on my machine (~0.99s in one JSON run), and `go test ./web/api/v1 -count=5` passes in 12.257s. I'll keep the PR limited to this test and include before/after timings.

**Comment by ogulcanaydogan** [comment]
Will add `t.Parallel()` to `discovery/uyuni` — continuing the same pattern.

**Comment by amarkdotdev** [comment]
Taking up `notifier/` — PR opened: #18922. All 26 top-level tests across `manager_test.go`, `alertmanager_test.go`, and `sendloop_test.go` parallelized.

**Comment by sahitya-chandra** [comment]
Hi @machine424, the whole `tracing/` package runs in 13ms total, each test rounds to 0.00s individually. So, I think it should be marked "Unnecessary" on the issue checklist, same as `config/` and `model/`
